File: algorithms/graph/dfs/DFS.cpp
#include <vector>

using namespace std;

void dfs(const vector<vector<int>> &graph, int node, vector<bool> &visited,
         vector<int> &traversal) {
  visited[node] = true;
  traversal.push_back(node);

  for (int neighbor : graph[node]) {
    if (!visited[neighbor]) {
      dfs(graph, neighbor, visited, traversal);
    }
  }
}

vector<int> depthFirstSearch(const vector<vector<int>> &graph, int start) {
  vector<int> traversal;
  vector<bool> visited(graph.size(), false);

  dfs(graph, start, visited, traversal);
  return traversal;
}

File: algorithms/graph/dfs/DFS.go
package main

import "fmt"

func dfs(graph map[string][]string, start string, visited map[string]bool) {
	visited[start] = true // mark current node as visited

	fmt.Println(start)

	for _, neighbour := range graph[start] {
		if !visited[neighbour] {
			dfs(graph, neighbour, visited)
		}
	}
}

func main() {
	graph := map[string][]string{
		"A": {"B", "C"},
		"B": {"D", "E"},
		"C": {"F"},
		"D": {},
		"E": {"F"},
		"F": {},
	}

	visited := make(map[string]bool)
	dfs(graph, "A", visited)
}
